controller/controller-common/src/main/java/com/pinterest/rocksplicator/controller/mysql/MySQLClusterManager.java

```diff
@@ -16,45 +16,154 @@
 
 package com.pinterest.rocksplicator.controller.mysql;
 
+import com.google.common.base.Strings;
 import com.pinterest.rocksplicator.controller.Cluster;
 import com.pinterest.rocksplicator.controller.ClusterManager;
 import com.pinterest.rocksplicator.controller.bean.HostBean;
+import com.pinterest.rocksplicator.controller.mysql.entity.TagEntity;
+import com.pinterest.rocksplicator.controller.mysql.entity.TagHostsEntity;
+import com.pinterest.rocksplicator.controller.mysql.entity.TagId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
-import java.util.Collections;
+import javax.persistence.Query;
+import java.util.HashSet;
 import java.util.List;
+import java.util.Set;
 
 /**
  * @author shu (shu@pinterest.com)
  */
 public class MySQLClusterManager extends MySQLBase implements ClusterManager{
 
+  private static final Logger LOG = LoggerFactory.getLogger(MySQLClusterManager.class);
+
   public MySQLClusterManager(String jdbcUrl, String dbUser, String dbPassword) {
     super(jdbcUrl, dbUser, dbPassword);
   }
 
+  private Set<HostBean> fromHostsString(final String hostsString) {
+    Set<HostBean> hostBeanList = new HashSet<>();
+    if (Strings.isNullOrEmpty(hostsString)) {
+      return hostBeanList;
+    }
+    String[] hostsArray = hostsString.split("#");
+    for (String hostString : hostsArray) {
+      hostBeanList.add(HostBean.fromString(hostString));
+    }
+    return hostBeanList;
+  }
+
+  private String toHostsString(Set<HostBean> hostBeans) {
+    if (hostBeans.isEmpty()) return "";
+    Set<String> hostStrings = new HashSet<String>();
+    for (HostBean hostBean : hostBeans) {
+      hostStrings.add(hostBean.toString());
+    }
+    return String.join("#", hostStrings);
+  }
+
+  private TagHostsEntity findTagHostsEntity(final Cluster cluster) {
+    Query query = getEntityManager().createNamedQuery("tag_hosts.getCluster")
+        .setParameter("namespace", cluster.getNamespace())
+        .setParameter("name", cluster.getName());
+    List<TagHostsEntity> entities = query.getResultList();
+    if (entities.isEmpty()) return null;
+    return entities.get(0);
+  }
+
+  private boolean persistTagHostsEntity(TagHostsEntity tagHostsEntity) {
+    beginTransaction();
+    getEntityManager().persist(tagHostsEntity);
+    getEntityManager().getTransaction().commit();
+    return true;
+  }
+
+  @Override
+  public boolean createCluster(final Cluster cluster) {
```

Comment:
is this operation protected by cluster lock? if not, i think we need to beginTransaction() from the beginning.

Reply:
good catch.
for other operations like register / unregister and blacklist, i plan to implement them in tasks, and task will automatically hold the cluster lock so they should be fine.